src/server.ts

```diff
@@ -136,121 +95,79 @@ export class Server {
+             * 3. Awaits for each instantiator that has a reader for that channel to connect
+             * 4. Notify writing runner to start sending data
+             * 5. Forwards received chunks to all registered receivers
+             * 6. For each chunk forward the processed control message back to the sending runner
+             * 7. Cleans up resources when the stream ends
              */
             sendStreamMessage: async (
-                stream: grpc.ServerDuplexStream<DataChunk, Id>,
+                stream: grpc.ServerDuplexStream<
+                    StreamChunk,
+                    ReceivingStreamControl
+                >,
             ) => {
-                const id = this.streamMsgId
-                this.streamMsgId += 1
-                this.logger.debug('Openin stream with id ' + id)
-
-                const obj: OpenStream = {
-                    done: [],
-                    receivers: [],
-                }
-                this.msgStreams[id] = obj
-
-                // Sending only message on the stream
-                await new Promise((res) => stream.write({ id: id }, res))
-
-                try {
-                    for await (const chunk of stream) {
-                        const c: DataChunk = chunk
-                        this.logger.debug('got chunk for stream ' + id)
-                        obj.done.push(c)
-                        for (const listener of obj.receivers) {
-                            await listener.data(c)
-                        }
-                    }
-                } catch (ex) {
-                    if (ex instanceof Error) {
-                        this.logger.debug(
-                            'Stream message stream closed:' +
-                                ex.name +
-                                ' ' +
-                                ex.message,
-                        )
-                    }
-                }
+                // Get the actual identifier of the channel
+                const identify = <StreamChunk>(
+                    await new Promise((res) => stream.once('data', res))
+                )
 
-                let c = obj.receivers.pop()
-                while (c !== undefined) {
-                    await c.close()
-                    c = obj.receivers.pop()
+                if (!identify.id) {
+                    throw 'no, expected first a stream identifier'
                 }
 
-                this.logger.debug(
-                    'data stream is finished, closing in 500ms ' + id,
-                )
-                setTimeout(() => {
-                    this.logger.debug('closing data stream ' + id)
-                    delete this.msgStreams[id]
-                    for (const receiver of obj.receivers) {
-                        receiver.close()
-                    }
-                }, 500)
+                this.orchestrator.startStreamMessage(identify.id, stream)
```

Comment:
promise is not awaited, is this intended?

Reply:
I'm not sure :(